lib/puppet/util/windows/sid.rb

```diff
@@ -97,6 +97,29 @@ def octet_string_to_sid_object(bytes)
     end
     module_function :octet_string_to_sid_object
 
+    def ads_to_sid_object(ads_object)
+      if !ads_object || !ads_object.respond_to?(:ole_respond_to?) ||
+        !ads_object.ole_respond_to?(:objectSID) || !ads_object.ole_respond_to?(:Name)
+        raise Puppet::Util::Windows::Error.new("ads_object must be an IAdsUser or IAdsGroup instance")
+      end
+      octet_string_to_sid_object(ads_object.objectSID)
+    rescue Puppet::Util::Windows::Error => e
+      # SID could not be mapped but is a valid SID representation
+      unresolvable = e.code == ERROR_NONE_MAPPED && valid_sid?(ads_object.Name)
+
+      raise if !unresolvable
+
+      Principal.new(
+        ads_object.Name + " (unresolvable)", # account
```

Comment:
@puppetlabs/windows / @joshcooper any thoughts on emitting this additional bit in the string for SIDs that can't be translated? IMHO, it's useful in the `puppet resource` output, but I can understand why it also might be considered extraneous.